common/persistence/persistenceErrorInjectionClients.go

```diff
@@ -0,0 +1,1686 @@
+	var persistenceErr error
+	if shouldForwardCallToPersistence(fakeErr) {
+		response, persistenceErr = p.persistence.EnqueueMessageToDLQ(ctx, message)
+	}
+
+	if fakeErr != nil {
+		return emptyMessageID, fakeErr
+	}
+	return response, persistenceErr
+}
+
+func (p *queueErrorInjectionPersistenceClient) ReadMessagesFromDLQ(
+	ctx context.Context,
+	firstMessageID int64,
+	lastMessageID int64,
+	pageSize int,
+	pageToken []byte,
+) ([]*QueueMessage, []byte, error) {
+	fakeErr := generateFakeError(p.errorRate)
+
+	var response []*QueueMessage
+	var token []byte
+	var persistenceErr error
+	if shouldForwardCallToPersistence(fakeErr) {
+		response, token, persistenceErr = p.persistence.ReadMessagesFromDLQ(ctx, firstMessageID, lastMessageID, pageSize, pageToken)
+	}
+
+	if fakeErr != nil {
+		return nil, nil, fakeErr
+	}
+	return response, token, persistenceErr
+}
+
+func (p *queueErrorInjectionPersistenceClient) RangeDeleteMessagesFromDLQ(
+	ctx context.Context,
+	firstMessageID int64,
+	lastMessageID int64,
+) error {
+	fakeErr := generateFakeError(p.errorRate)
+
+	var persistenceErr error
+	if shouldForwardCallToPersistence(fakeErr) {
+		persistenceErr = p.persistence.RangeDeleteMessagesFromDLQ(ctx, firstMessageID, lastMessageID)
+	}
+
+	if fakeErr != nil {
+		p.logger.Error(msgInjectedFakeErr, tag.Error(fakeErr))
+		return fakeErr
+	}
+	return persistenceErr
+}
+
+func (p *queueErrorInjectionPersistenceClient) UpdateDLQAckLevel(
+	ctx context.Context,
+	messageID int64,
+	clusterName string,
+) error {
+	fakeErr := generateFakeError(p.errorRate)
+
+	var persistenceErr error
+	if shouldForwardCallToPersistence(fakeErr) {
+		persistenceErr = p.persistence.UpdateDLQAckLevel(ctx, messageID, clusterName)
+	}
+
+	if fakeErr != nil {
+		p.logger.Error(msgInjectedFakeErr, tag.Error(fakeErr))
+		return fakeErr
+	}
+	return persistenceErr
+}
+
+func (p *queueErrorInjectionPersistenceClient) GetDLQAckLevels(
+	ctx context.Context,
+) (map[string]int64, error) {
+	fakeErr := generateFakeError(p.errorRate)
+
+	var response map[string]int64
+	var persistenceErr error
+	if shouldForwardCallToPersistence(fakeErr) {
+		response, persistenceErr = p.persistence.GetDLQAckLevels(ctx)
+	}
+
+	if fakeErr != nil {
+		return nil, fakeErr
+	}
+	return response, persistenceErr
+}
+
+func (p *queueErrorInjectionPersistenceClient) DeleteMessageFromDLQ(
+	ctx context.Context,
+	messageID int64,
+) error {
+	fakeErr := generateFakeError(p.errorRate)
+
+	var persistenceErr error
+	if shouldForwardCallToPersistence(fakeErr) {
+		persistenceErr = p.persistence.DeleteMessageFromDLQ(ctx, messageID)
+	}
+
+	if fakeErr != nil {
+		p.logger.Error(msgInjectedFakeErr, tag.Error(fakeErr))
+		return fakeErr
+	}
+	return persistenceErr
+}
+
+func (p *queueErrorInjectionPersistenceClient) Close() {
+	p.persistence.Close()
+}
+
+func shouldForwardCallToPersistence(
+	err error,
+) bool {
+	return err == nil || err == ErrFakeTimeout || err == ErrFakeUnhandled
+}
+
+func generateFakeError(
```

Comment:
Different APIs could return some different errors. Should we also consider this? Currently, it returns some generic errors. Will we consider to expand the error type to workflowalreadystarted/conditionupdatefailure etc.?

Reply:
I went through all the error types we defined and I don't think we can randomly return those errors (like entity not exist, workflow already started, domain not active, conditional update failed, etc) without checking the actual state in DB. Those errors are typically non-retryable and will cause the load we are running to fail.